### ISSUE TITLE
Finding a text in the page

### ISSUE THREAD

**Issue by mlajtos** [mlajtos/mosaic#8]
# Motivation

Finding specific text in the page is a must-have feature. Without it, one can't use Mosaic on a daily basis, hence can't take advantage of the tiling in every situation.

# Design

## Proposal A

Search bar for a specific text in a page can be integrated into the omnibar. [!bangs](https://duckduckgo.com/bang) are an extremely good abstraction for the search queries, and searching text in a page fits this category. Having a `!page` or `!inpage` pseudo bang would trigger a special mode for omnibar (or toolbar) that would add necessary buttons (previous match, next match).

## Proposal B

Copy and refine existing design from prior art.

# Prior art

## Firefox

Searchbar takes full width of the view below the website.

![disabling-Firefox-find-bar-page-dimming-and-highlight-all-features](https://user-images.githubusercontent.com/1047512/90039541-0fce4380-dcc7-11ea-9960-b5e4c1932988.png)
(source: https://techdows.com/2016/09/disable-firefox-find-bar-page-dimming-highlight-all-features.html)

## Safari

Search bar takes full width of the view and pushes the website down.

![Find_on_page_bar_and_matches_on_safari_browser_mac_computer](https://user-images.githubusercontent.com/1047512/90038604-e7921500-dcc5-11ea-88f9-79eced4c9018.png)
(source: https://browserhow.com/how-to-search-text-and-find-on-page-in-safari-mac/)

## Chrome

Search bar is compact and floating above the page.

![browser-will-find-this-word-on-page](https://user-images.githubusercontent.com/1047512/90038912-435c9e00-dcc6-11ea-83fd-2a022a12b2aa.jpg)
(source: https://rocketfiles.com/articles/how-to-search-for-a-word-on-a-web-page-chrome)

## Vivaldi

Search bar takes whole width of the view and is aligned to the center.

<img width="960" alt="Find-in-Page-example" src="https://user-images.githubusercontent.com/1047512/90039017-638c5d00-dcc6-11ea-9ca3-24befaaf6732.png">

(source: https://rocketfiles.com/articles/how-to-search-for-a-word-on-a-web-page-chrome)

# Relevant documentation
- event [`found-in-page`](https://github.com/electron/electron/blob/master/docs/api/web-contents.md#event-found-in-page)
- method [`findInPage`](https://github.com/electron/electron/blob/master/docs/api/web-contents.md#contentsfindinpagetext-options)
- method [`stopFindInPage`](https://github.com/electron/electron/blob/master/docs/api/web-contents.md#contentsstopfindinpageaction)

**Comment by marc2332** [comment]
I am working on this in https://github.com/marc2332/mosaic.

So far:
* Ctrl+F opens a floating search box ( that's my design, it can be obviously discussed)
* Esc closes it
* The floating box has a input where you can write and a button to close next to it


Result:

![imagen](https://user-images.githubusercontent.com/38158676/90134594-b2430100-dd71-11ea-8550-e349d78ffa3b.png)

![imagen](https://user-images.githubusercontent.com/38158676/90134638-bf5ff000-dd71-11ea-89de-29a018b9d411.png)

Ideas:
* Added arrows to search Up/Down
* Add an option to toggle case sensitive

**Note**: I have never used recoil, but I am trying my best 👍🏻

**Comment by michalklempa** [comment]
May I suggest some key binding to 'jump to next search result' and 'jump to previous?', what are the usual bindings on these (not `n` and `N` as in `less` :D)

**Comment by marc2332** [comment]
Yep! 

In Firefox these shortcuts are (pretty bad in my opinion):
Ctrl+G -> Next result
Shift F3 -> Previous result

I would definitely use ones more friendly, specially for the 'previous' one. And also add some Up and Down arrows to do the same

**Comment by mlajtos** [comment]
Hey @marc2332,

I tried your Find in Page implementation and it works quite okay. Go ahead and make a PR.

Don't worry about Recoil.js, I jumped on it because I wanted to try out something new, so we are in the same boat being beginners. 🙂  So far I can see it has shortcomings (especially for this project 😅), but I hope they will be worked out as we go further.

About keybindings – I use <kbd>Enter</kbd> for next and <kbd>Shift-Enter</kbd> for previous. These only makes sense. At least for me.

Btw nice touch adding animation. 🙂
